### DISCUSSION TITLE
delay / timeout

### DISCUSSION THREAD

**Discussion by avarama** [purpleteam-labs/purpleteam#118] [General]
Is there a way to slow down the underlying selenium interactions with the web app? I guess it is designed to wait for elements, but I'm suspecting purpleteam is timing out before my elements are drawn. Is there a configuration for the delay between steps, or any directive to specify specific waits? Thanks!

**Comment by binarymist** [comment]
Hi @avarama.

Did you try [viewing the Selenium browser](https://github.com/purpleteam-labs/purpleteam-s2-containers#viewing-browser-in-selenium-container) to see what's actually happening?

Yes there are waits.

Next step after just seeing what's happening with the Selenium browser would probably be to [debug](https://purpleteam-labs.com/doc/local/workflow/#app-scanner-and-sub-processes) the Cucumber sub-process.

Les us know how that goes or if you need any more pointers?

**Reply by avarama** [reply]
I'm running the tool in an ubuntu server. I'd love to avoid having to install VSCode or any GUI stuff, as I am running with a low end AWS instance. Is there a global timeout or a way to instruct a wait in the Job file? Otherwise I'd need to install xorg or other X server in order to run VSCode in the remote server for debugging. Am I right?

**Comment by binarymist** [comment]
Hmm, that could be a problem. There's no global wait timeout for webdriver or waits specified in _Job_ file. Any waits should be be in: https://github.com/purpleteam-labs/purpleteam-app-scanner/blob/main/src/clients/browser.js

**Reply by avarama** [reply]
Do you think I could stuff some clever JS on the presenter's `beforeRequest` hook to introduce a delay in the events loop? Here: https://github.com/purpleteam-labs/purpleteam/blob/main/src/presenter/apiDecoratingAdapter.js#L86

**Reply by binarymist** [reply]
That would not affect Server Sent Events. It would only affect standard http(s) requests. It also has nothing to do with the selenium interactions in the app-scanner.

It sounds like you're proposing to add the ability to supply configurable webdriver waits from the CLI? Which could be worth doing. It'd mean you'd need to have a development environment set-up anyway. If this is the case, then you'll be able to find out what the actual problem is... before proposing change.

Thoughts?

**Reply by avarama** [reply]
Makes total sense. I think I'll try to do a remote debugging session, with vscode in my local computer and purpleteam running in the remote server, which has no X server running.

**Reply by binarymist** [reply]
Hopefully you can view VNC into the remote selenium browser via SSH
